File: resources/oldMatlabFiles/SimParam.m
classdef SimParam
    %SIMULATION Summary of this class goes here
    %   Detailed explanation goes here
    
    properties
        dt;            % Simulation time-step
        timeForcing;   % Weather data time-step
        month;         % Begin month
        day;           % Begin day of the month
        days;          % Number of days of simulation
        timePrint;     % time-step for printing outputs
        timeDay;       % number of timesteps in a design-day
        timeSim;       % number of timesteps of the simulation
        timeMax;       % total simulation time (s)
        nt;            % total number of timesteps
        timeFinal;     % final timestep of simulation
        timeInitial;   % initial timestep of simulation
        secDay;        % seconds of one day (s)
        hourDay;       % hour of the day (0 - 23hr)
        inobis;        % julian day at the end of each month
        julian;        % julian day
    end
    
    methods
        function obj = SimParam(dt,timefor,M,DAY,days)
            % class constructor
            if(nargin > 0)
                obj.dt = dt;
                obj.timeForcing = timefor;
                obj.month = M;
                obj.day = DAY;
                obj.days = days;
                obj.timePrint = timefor;
                obj.timeDay = 24*3600/timefor;
                obj.timeSim = obj.timeDay*days;
                obj.timeMax = 24.*3600.*days;
                obj.nt = fix(obj.timeMax/obj.dt+1);
                obj.inobis = [0,31,59,90,120,151,181,212,243,273,304,334];
                obj.julian = obj.inobis(M)+DAY-1;
                H1 = (obj.inobis(M)+DAY-1)*obj.timeDay;
                obj.timeInitial = H1 + 8;
                obj.timeFinal = H1 + obj.timeDay*obj.days - 1 + 8;
                obj.secDay = 0;
                obj.hourDay = 0;
            end
        end
        
        % update date
        function obj = UpdateDate(obj)
            obj.secDay = obj.secDay + obj.dt;
            if obj.secDay == 3600*24
                obj.day = obj.day+1;
                obj.julian = obj.julian+1;
                obj.secDay = 0;
                for j = 1:12
                    if obj.julian == obj.inobis(j)
                        obj.month = obj.month + 1;
                        obj.day = 1;
                    end
                end
            end
            obj.hourDay = floor(obj.secDay/3600);       % 0 - 23hr
        end
    end
end

File: uwg/simparam.py
"""Class for simulation time parameters."""
from __future__ import division, print_function

try:
    range = xrange
except NameError:
    pass

import math
from .utilities import is_near_zero


class SimParam(object):
    """Calculates simulation time parameters.

    Simulation time parameters are calculated based on initial date and weather
    data time step, and simulation timestep.

    Args:
        dt: Number for simulation time step in seconds.
        timefor: Number for weather data time-step in seconds.
        M: Number between 1 and 12 indicating the simulation start month.
        DAY: Number between 1 and 31 indicating the simulation start day.
        days: Number indicating number of days to simulate.

    Properties:
        * dt -- uwg time simulation time step
        * timeForcing -- weather data timestep
        * month -- start month
        * day -- start day
        * days -- number of days in simulation
        * timePrint -- weather data timestep
        * timeDay -- number of times weather senses in a day
        * timeSim -- number of steps in weather data simulation
        * timeMax -- total seconds in simulation days
        * nt -- total number of timesteps
        * inobis -- list of julian dates for the start of the months
        * julian -- simulation start julian date
        * timeInitial -- epw sensor data for initial time based on julian day & timesteps
        * timeFinal -- epw sensor data for final time based on julian day & timesteps
        * secDay -- current seconds in day
        * hourDay -- current hours in day
    """
    TIMESTEP_CONFLICT_MSG = "TIMESTEP ERROR! Timestep must be a factor of 3600."

    def __init__(self, dt, timefor, M, DAY, days):
        self.dt = dt
        self.timeForcing = timefor
        self.month = int(M)
        self.day = DAY
        self.days = days
        self.timePrint = timefor
        self.timeDay = 24 * 3600 / timefor
        self.timeSim = self.timeDay * days
        self.timeMax = 24. * 3600. * days
        self.nt = int(round(self.timeMax / self.dt + 1))
        self.inobis = [0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334]
        self.julian = self.inobis[self.month - 1] + DAY - 1
        # H1: (julian day * number of timesteps in a day) == sensor data index in epw
        H1 = int((self.inobis[self.month - 1] + DAY - 1) * self.timeDay)
        self.timeInitial = H1 + 8
        self.timeFinal = int(H1 + self.timeDay * self.days - 1 + 8)
        self.secDay = 0
        self.hourDay = 0

    def update_date(self):
        self.secDay = self.secDay + self.dt

        if is_near_zero(self.secDay - (3600 * 24)):
            self.day += 1
            self.julian = self.julian + 1
            self.secDay = 0.
            for j in range(12):
                if is_near_zero(self.julian - self.inobis[j]):
                    self.month = self.month + 1
                    self.day = 1

        if self.secDay > (3600 * 24):
            raise Exception("{}. CURRENTLY AT {}.".format(
                self.TIMESTEP_CONFLICT_MSG, self.dt))

        self.hourDay = int(math.floor(self.secDay / 3600.))  # 0 - 23hr

    def __repr__(self):
        return "SimParam: Start = {}/{}, days = {}, timestep = {}s".format(
            self.month, int(self.day), self.days, self.dt)
